File: functions-cart-checkout-validation-js/src/cart_validations_generate_run.ts
import type {
  CartValidationsGenerateRunInput,
  CartValidationsGenerateRunResult,
  FunctionError,
} from "../generated/api";

export function cart_validations_generate_run(input: CartValidationsGenerateRunInput): CartValidationsGenerateRunResult {
  const errors: FunctionError[] = input.cart.lines
    .filter(({ quantity }) => quantity > 1)
    .map(() => ({
      localizedMessage: "Not possible to order more than one of each",
      target: "$.cart",
    }));

  return {
    errors
  }
};

File: functions-cart-checkout-validation-rs/src/cart_validations_generate_run.rs
use super::schema;
use shopify_function::prelude::*;
use shopify_function::Result;

#[shopify_function]
fn cart_validations_generate_run(
    input: schema::cart_validations_generate_run::Input,
) -> Result<schema::CartValidationsGenerateRunResult> {
    let mut operations = Vec::new();
    let mut errors = Vec::new();

    if input
        .cart()
        .lines()
        .iter()
        .map(|line| *line.quantity())
        .any(|quantity| quantity > 1)
    {
        errors.push(schema::ValidationError {
            message: "Not possible to order more than one of each".to_owned(),
            target: "$.cart".to_owned(),
        })
    }
    let operation = schema::ValidationAddOperation { errors };
    operations.push(schema::Operation::ValidationAdd(operation));

    Ok(schema::CartValidationsGenerateRunResult { operations })
}
